keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,201 @@
+- https://github.com/kubernetes/kubernetes/issues/24238
+- https://github.com/kubernetes/kubernetes/issues/34292
+- https://github.com/kubernetes/kubernetes/issues/6979
+- https://github.com/kubernetes/kubernetes/issues/33766
+- https://github.com/kubernetes/kubernetes/issues/24198
+- https://github.com/kubernetes/kubernetes/issues/60340
+
+And then, for other people:
+- https://github.com/rancher/rancher/issues/13584
+- https://github.com/helm/charts/pull/12319
+- https://github.com/EnMasseProject/enmasse/pull/1974
+- https://github.com/helm/charts/pull/11546
+- https://github.com/kubernetes/kubernetes/pull/35343
+
+### Goals
+
+The goal is to enable a union or "oneof" semantic in Kubernetes types, both for
+in-tree types and for CRDs.
+
+### Non-Goals
+
+We're not planning to use this KEP to release the feature, but mostly as a way
+to document what we're doing.
+
+## Proposal
+
+In order to support unions in a backward compatible way in kubernetes, we're
+proposing the following changes.
+
+### Go tags
+
+We're proposing a new type of tags for go types (in-tree types, and also
+kubebuilder types):
+
+- `// +oneof` before a field means that it is part of the union. Multiple fields
+  can have this prefix. Alternatively, the form `// +oneof=<name>` allows to
+  provide the discriminated name for that specific field. These fields MUST BE
+  optional, omitempty and pointers.
+- `// +discriminator` before a field means that this field is the discriminator
+  for the union. Only one field per structure can have this prefix. This field HAS
+  TO be a string, and CAN be optional.
+
+Note that this means that only one "union" type can be defined per structure,
+which is a deliberate decision.
+
+### OpenAPI
+
+OpenAPI v3 already allows a "oneOf" form, which is accepted by CRD validation
+(and will continue to be accepted in the future), they will be interpreted
+properly if possible (see CRD GA effort).
+
+Since OpenAPI v2 has no support for oneOf, a couple of kubernetes specific
+extensions will be added to OpenAPI:
+- `x-kubernetes-oneof-discriminator: true` is set for the discriminator field,
+- `x-kubernetes-oneof: "<discriminatedName>` is set for discriminated fields,
+
+As for Go types, we only accept one union per structure, which means there can
+be only one discriminator, and all other `oneof` fields are assumed to belong to
+the same union.
+
+### Discriminator
+
+For backward compatibility reasons, discriminators can be added to existing
+union structures as an optional string. This has a nice property that it's going
+to allow conflict detection when a union field is flipped.
+
+The value of the discriminator is going to be set automatically by the apiserver
+when a new field is changed in the union, and will be interpreted as an
+intention to clear all the other fields when its value is changed. See section
+below.
+
+### Normalization process
+
+The normalization process (which runs before validating that the union property
```

Comment:
Is this going to be handled automatically as part of the REST strategy or something?  E.g. are we generating "normalize" functions that get called before validation?  It would be amazing if this was magical and Just Worked without individual API authors having to do anything.

Reply:
I'll clarify that but basically: since we now have access to the openapi in all the handler path (CREATE/UPDATE/PATCH/"Apply") to track fields manager, we can also automatically process these normalization before validation.